diet.1
.TH diet 1 "April 2001"
.SH NAME
.PP
diet \- mangle gcc command line arguments
.SH SYNOPSIS
.PP
.B diet [-v] [-Os]
[\fInormal gcc command line\fP]
.SH DESCRIPTION
.PP
diet is a small wrapper around gcc or clang.  It will try to look at the command
line you specify and try to mangle it so that gcc/clang will compile against
the diet libc header files and link against the diet libc itself.

diet tries to be smart for cross compiling.  If the first argument is
not gcc but sparc-linux-gcc, diet will guess that you want to cross
compile for sparc and use bin-sparc/dietlibc.a instead of the dietlibc.a
for your default architecture. This does not work yet for clang.

The -v option will make diet print the modified gcc command line before
executing it.

When passed the -Os option before the gcc argument, diet will mangle the
gcc options to include the best known for-size optimization settings for
the platform.
.SH FILES
~/.diet/\fIcompiler\fR may contain compiler options as you would
specify them on the command line, i.e. separated by spaces.  Those will
then be used instead of the built-in defaults for diet -Os.
.SH AUTHOR
Felix von Leitner <felix-dietlibc@fefe.de>


CHANGES
0.34:
  move stpcpy from libcompat to lib as it is now part of POSIX (Enrico Scholz)
  add linux/seccomp.h linux/prctl.h linux/audit.h linux/elf-em.h linux/filter.h
  SECURITY: make sure all assembler files mark the stack non-executable so
    the kernel can map the stack of the whole program as non-executable
  support compiling with clang (no cross compiling yet, and the resulting
    binary code is significantly larger than with gcc)
  implement experimental C11 threads
  fix fgets (Andreas Förster)
  add eventfd, inotify_init1, epoll_init1, accept4, recvmmsg, sendmmsg
  add dup3, fanotify_init, fanotify_mark, signalfd4, timerfd_settime
  add timerfd_gettime, setns, memfd_create, getrandom, syncfs
  add preadv, pwritev
  add getcontext, setcontext, makecontext, swapcontext (i386 and x86_64 only for now)
  add implementation of hsearch and friends
  add strverscmp (like GNU version but without special treatment for leading zeros)
  fix cuserid
  SECURITY: fix integer overflow in scandir/scandir64 if your selector
    leaves more than 1 billion filenames
  fix memory leaks in out of memory error paths in regcomp and librpc
    udp and tcp server code (found by the llvm static analyzer)
  fix returning more than one match in regexec (Stefan Wagner)
  add getentropy, lazy arc4random clone (OpenBSD)
  define uintptr_t using __UINTPTR_TYPE__ (32-bit platforms may need a make clean)

0.33:
  first stab at getting TLS to work in actual threads
  use fs: to find current thread on x86_64
  add fallocate, posix_fallocate
  work around Ubuntu who added -fstack-protector to their gcc, which
    added stack protector code to the function from the dietlibc startup
    code that sets up the registers and segments used by the stack
    protector mechanism.
  fix sys/personality.h (Mathias Krause)
  update sys/ptrace.h (Nikola Vladov)
  fix fork+pthread closing fd 0 (Tim München)
  finally add stdout tty detection and turn off linewise buffering if !tty
  enlarge dev_t to 32-bit on i386, arm, sparc, s390 (Pascal Terjan)
  update sys/stat.h to include nsec and enlarge dev_t
  add WANT_VALGRIND_SUPPORT in dietfeatures.h.  When #defined, the
    startup code will check if the program is running under valgrind,
    and if it is, the SIMD optimized string routines will switch back to
    a slower version that does not generate valgrind false positives.
  add WANT_LLMNR in dietfeatures.h.  Now you can do "ping windowsbox"
    and it works if that box runs win7 and trusts the LAN; the down side
    is that LLMNR does not like the .local prefix, so DNS does zeroconf
    lookup for unqualified names now, too.
  add WANT_GNU_STARTUP_BLOAT in dietfeatures.h to support
    program_invocation_short_name and program_invocation_name
  add INT*_MIN, INT*_MAX, UINT*_MAX, etc to stdint.h
  remove const from stdsignal prototype to agree with standard
  only count %[] conversions in *scanf if they matched at least one char
    (Indan Zupancic)
  make sure __sc_nr_cpus does not dump core if /proc/cpuinfo is larger
    than 2048 bytes (Bela Lubkin)
  use vsyscall for time on x86_64 instead of wrapping gettimeofday (Nikola Vladov)
  CPU affinity support (Indan Zupancic)
  lots of ARM fixes (Enrico Scholz)
  make sure strftime does not write more than max bytes when writing \0
    (Tobias S. Josefowitz)
  change gettimeofday vsyscall to use the vdso
  add WANT_HOSTS_GIVEUP_EARLY in dietfeatures.h
  various build fixes, add sys/atomic.h and the __CAS function
  redo nice with get/setpriority as glibc does (Thierry Vignaud)
  add fdopendir
  add openat and friends
  srand(0) made rand() return a series of 0 instead of random data
    (https://banu.com/blog/42/openbsd-bug-in-the-random-function/)
  add SO_RCVBUFFORCE and SO_SNDBUFFORCE
  first stab at x32 support

0.32:
  add mbsinit, mbrlen, mbtowc, wctomb, mbrtowc, wcrtomb, wcwidth
  add btowc, wctype, iswctype
  add some wide char routines
  add statvfs, fstatvfs
  add some amd64 asm string routines
  add some constants for inotify
  several fixes from Enrico Scholz (printf, string routines on big endian)
  add some more syscalls
  some minor cleanups
  shrink some string functions (Kris Katterjohn)
  prepare x86_64 for syscall numbers > 255
  fix unaligned memcpy if WANT_SMALL_STRING_ROUTINES is not defined
    (Enrico Scholz)
  fix utmp, add utmpx support (R.L. Horn)
  fix the dynamic library for the ARM architecture (Gernot Tenchio)
  get ARM port to work with EABI (Tristan Newby)
  fix of scanf with negative floats (Andre Oliveira)
  use vsyscall to do time(2) on x86-64
  implement back references in basic regexes
  make regex faster (coalesce n CHAR matches to one STRING match)
  stdio speedup
  use randomness passed in ELF auxvec instead of /dev/urandom in startup
    (strace looks cleaner now)
  use ELF auxvec to implement getpagesize
  support thread local storage (w00t!!!!)
  don't crash in getaddrinfo if hints is NULL (Gernot Tenchio)
  add -D_REENTRANT for -pthread in diet
  don't segfault in setlocale if locale is NULL (Nikola Vladov)
  implement _SC_PHYS_PAGES in sysconf()
  add some gcc attributes and restrict keywords to string.h

0.31:
  getopt_long and getopt_long_only now refuse ambiguous abbreviations
  fix printf %u/%x on 64-bit platforms
  fix fd leak in if_nameindex (Florian Westphal)
  fix printf %g (well, get closer)
  fix sparc64 longjmp (Rene Rebe)
  add dn_skipname
  add pthread prototypes that are needed to compile libgcc when
    compiling a cross compiler gcc
  fix ftw64
  fixes for some corner cases from Ilja van Sprundel:
    - tempnam
    - strsignal
    - strftime
    - sunrpc code
  faster qsort
  support %j in printf (Gernot Tenchio)
  do omit the frame pointer on x86_64
  fix pthread_cond_timedwait (Gernot Tenchio)
  fix getcwd, ftw, ftw64, dirname (Ilja van Sprundel)
    Some fd leaks in error conditions.
  fix stackguard to be .quad instead of .long (cause subtle memory

diet.c
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <write12.h>

#include "dietfeatures.h"

/* goal:
 *   when invoked as
 * "diet gcc -c t.c"
 *   exec
 * "gcc -I/path/to/dietlibc/include -c t.c"
 *
 *   when invoked as
 * "diet sparc-linux-gcc -o t t.o"
 *   exec
 * "sparc-linux-gcc -nostdlib -static -o t t.o /path/to/dietlibc/bin-sparc/start.o /path/to/dietlibc/bin-sparc/dietlibc.a"
*/

static void error(const char *message) {
  __write2(message);
  exit(1);
}

static const char* Os[] = {
  "i386","-Os","-mpreferred-stack-boundary=4",
	 "-falign-functions=1","-falign-jumps=1",
	 "-falign-loops=1","-fomit-frame-pointer",0,
  "x86_64","-Os",0,
  "x32","-Os",0,
  "sparc","-Os","-mcpu=supersparc",0,
  "sparc64","-Os","-m64","-mhard-quad-float",0,
  "alpha","-Os","-fomit-frame-pointer",0,
  "arm","-Os","-fomit-frame-pointer",0,
  "mips","-Os","-fomit-frame-pointer","-march=mips2",0,
  "mipsel","-Os","-fomit-frame-pointer","-march=mips2",0,
  "ppc","-Os","-fomit-frame-pointer","-mpowerpc-gpopt","-mpowerpc-gfxopt",0,
  "ppc64","-Os","-fomit-frame-pointer","-mpowerpc-gpopt","-mpowerpc-gfxopt",0,
  "s390","-Os","-fomit-frame-pointer",0,
  "s390x","-Os","-fomit-frame-pointer",0,
  "sh","-Os","-fomit-frame-pointer",0,
  "ia64","-Os","-fno-omit-frame-pointer",0,
  0};

static void usage(void) {
  __write2(
#ifdef __DYN_LIB
	   "dyn-"
#endif
	   "diet version " VERSION
#ifndef INSTALLVERSION
	   " (non-install version in source tree)"
#endif
	   "\n\n");
  error("usage: diet [-v] [-Os] gcc command line\n"
	"e.g.   diet -Os gcc -c t.c\n"
	"or     diet sparc-linux-gcc -o foo foo.c bar.o\n");
}

int main(int argc,char *argv[]) {
  int _link=0;
  int compile=0;
  int preprocess=0;
  int verbose=0;
  int profile=0;
  char* diethome;
  char* platform;
#ifdef __DYN_LIB
  int shared=0;
#endif
  char* shortplatform=0;
#ifdef WANT_SAFEGUARD
  char safeguard1[]="-include";
  char* safeguard2;
#endif
  const char *nostdlib="-nostdlib";
  const char *libgcc="-lgcc";
  char *libpthread="-lpthread";
  char dashL[1000];
  char dashstatic[]="-static";
  int i;
  int mangleopts=0;
  int printpath=0;
  char manglebuf[1024];
  int m;

  if (!(diethome = getenv("DIETHOME")))
    diethome=DIETHOME;
#ifdef WANT_SAFEGUARD
  safeguard2=alloca(strlen(diethome)+30);
  strcpy(safeguard2, diethome);
  strcat(safeguard2, "/include/dietref.h");
#endif
  platform=alloca(strlen(diethome)+100);
  strcpy(platform,diethome);
#ifdef INSTALLVERSION
  strcat(platform,"/lib-");
#else
#ifndef __DYN_LIB
  strcat(platform,"/bin-");
#else
  strcat(platform,"/pic-");
#endif
#endif
  strcpy(dashL,"-L");

  do {
    if (!argv[1]) usage();
    if (!strcmp(argv[1],"-v")) {
      ++argv; --argc;
      verbose=1;
    } else if (!strcmp(argv[1],"-Os")) {
      ++argv; --argc;
      mangleopts=1;
    } else if (!strcmp(argv[1],"-L")) {
      ++argv; --argc;
      printpath=1;
    } else break;
  } while (1);
  {
    m=0;
    for (i=1; i<argc; ++i) {
      if (!strcmp(argv[i],"-m32")) m=32; else
      if (!strcmp(argv[i],"-mx32")) m=33; else
      if (!strcmp(argv[i],"-m64")) m=64;
    }
  }
  {
    char *cc=argv[1];
    char *tmp=strchr(cc,0)-2;
    char *tmp2,*tmp3;
    if (tmp<cc) goto donttouch;
    if (!strstr(cc,"cc") && !strstr(cc,"clang")) goto donttouch;
    if ((tmp2=strstr(cc,"linux-"))) {	/* cross compiling? */
      int len=strlen(platform);
      --tmp2;
      tmp3=strchr(cc,'-');
      if (tmp3<tmp2) tmp2=tmp3;
      if (tmp2-cc>90) error("platform name too long!\n");
      shortplatform=platform+len;
      memmove(shortplatform,argv[1],(size_t)(tmp2-cc));
      platform[tmp2-cc+len]=0;
      if (shortplatform[0]=='i' && shortplatform[2]=='8' && shortplatform[3]=='6') shortplatform[1]='3';
      else if (strcmp(shortplatform, "powerpc") == 0)
        strcpy(shortplatform, "ppc");
      else if (strcmp(shortplatform, "powerpc64") == 0)
        strcpy(shortplatform, "ppc64");
    } else {
#ifdef __sparc__
#ifdef __arch64__
      shortplatform="sparc64";
#else
      shortplatform="sparc";
#endif
#endif
#ifdef __powerpc__
      shortplatform="ppc";
#endif
#ifdef __powerpc64__
      shortplatform="ppc64";
#endif
#ifdef __i386__
      shortplatform="i386";
#endif
#ifdef __alpha__
      shortplatform="alpha";
#endif
#ifdef __arm__
      shortplatform="arm";
#endif
#ifdef __MIPSEL__
      shortplatform="mipsel";
#endif
#ifdef __MIPSEB__
      shortplatform="mips";
#endif
#ifdef __s390x__
      shortplatform="s390x";
#else
#ifdef __s390__
      shortplatform="s390";
#endif
#endif
#ifdef __sh__
      shortplatform="sh";
#endif
#ifdef __hppa__
      shortplatform="parisc";
#endif
#ifdef __x86_64__
      shortplatform=(m==32?"i386":(m==33?"x32":"x86_64"));
#endif
#ifdef __ia64__
      shortplatform="ia64";
#endif
      {
	char *tmp4=platform+strlen(platform);
	strcpy(tmp4,shortplatform);
	shortplatform=tmp4;
      }
    }
    /* MIPS needs special handling.  If argv contains -EL, change
     * platform name to mipsel */
    if (!strcmp(shortplatform,"mips")) {
      for (i=1; i<argc; ++i)
	if (!strcmp(argv[i],"-EL"))
	  strcpy(shortplatform,"mipsel");
    }
    if (printpath) {
      write(1,platform,strlen(platform));
      return 0;
    }
    strcat(dashL,platform);
    if (strcmp(tmp,"ld")) {
      char **newargv;
      char **dest;
      char *a,*b,*c;
#ifdef WANT_DYNAMIC
      char *d,*e,*f;
#endif
/* we need to add -I... if the command line contains -c, -S or -E */
      for (i=2; i<argc; ++i) {
	if (argv[i][0]=='-' && argv[i][1]=='M')
	  goto pp;
	if (!strcmp(argv[i],"-pg"))
	  profile=1;
	if (!strcmp(argv[i],"-c") || !strcmp(argv[i],"-S"))
	  compile=1;
	if (!strcmp(argv[i],"-E"))
pp:
	  preprocess=compile=1;
      }
/* we need to add -nostdlib if we are not compiling */
      _link=!compile;
#ifdef __DYN_LIB
      if (_link) {
	for (i=2; i<argc; ++i)
	  if (!strcmp(argv[i],"-shared")) {
	    shared=1;
	    _link=0;
	  }
      }
#endif
#if 0
      for (i=2; i<argc; ++i)
	if (!strcmp(argv[i],"-o"))
	  if (!compile) _link=1;
#endif

      newargv=alloca(sizeof(char*)*(argc+100));
      a=alloca(strlen(diethome)+20);
      b=alloca(strlen(platform)+20);
      c=alloca(strlen(platform)+20);

      strcpy(a,diethome); strcat(a,"/include");
#ifndef __DYN_LIB
      strcpy(b,platform);
      if (profile) strcat(b,"/pstart.o"); else strcat(b,"/start.o");
#ifdef INSTALLVERSION
      strcpy(c,platform); strcat(c,"/libc.a");
#else
      strcpy(c,platform); strcat(c,"/dietlibc.a");
#endif
#else
      strcpy(b,platform); strcat(b,"/dstart.o");
      strcpy(c,"-lc");
#endif

#ifdef WANT_DYNAMIC
      d=alloca(strlen(platform)+20);
      e=alloca(strlen(platform)+20);
#ifdef __DYN_LIB
      strcpy(d,platform);
      strcpy(e,platform);
      if (shared)
	strcat(d,"/dyn_so_start.o");
#ifdef INSTALLVERSION
      else
	strcat(d,"/dyn_dstart.o");
      strcat(e,"/dyn_dstop.o");
#else
      else
	strcat(d,"/dyn_start.o");
      strcat(e,"/dyn_stop.o");
#endif
#else
      strcpy(d,platform); strcat(d,"/dyn_start.o");
      strcpy(e,platform); strcat(e,"/dyn_stop.o");
#endif
#endif

      dest=newargv;
      *dest++=argv[1];
      if (argv[2]) {
	if (!strcmp(argv[2],"-V")) {
	  *dest++=argv[2];
	  *dest++=argv[3];
	  argv+=2;
	  argc-=2;
	} else if (!memcmp(argv[2],"-V",2)) {
	  *dest++=argv[2];
	  ++argv;
	  --argc;
	}
      }
#ifndef __DYN_LIB
      if (_link) { *dest++=(char*)nostdlib; *dest++=dashstatic; *dest++=dashL; }
#else
      /* avoid R_*_COPY relocations */
      *dest++="-fPIC";
      if (_link || shared) { *dest++=(char*)nostdlib; *dest++=dashL; }
#endif
#ifdef WANT_SAFEGUARD
      if (compile && !preprocess) {
	*dest++=safeguard1;
	*dest++=safeguard2;
      }
#endif
      if (_link) { *dest++=b; }
#ifdef WANT_DYNAMIC
      if (_link) { *dest++=d; }
#endif
      for (i=2; i<argc; ++i) {
	if (!strcmp(argv[i],"-pthread")) {
	  *dest++="-D_REENTRANT";
	  if (_link) *dest++="-lpthread";
	  continue;
	}
	if (mangleopts)
	  if (argv[i][0]=='-' && (argv[i][1]=='O' || argv[i][1]=='f' ||
				  (argv[i][1]=='m' && argv[i][2]!='3' && argv[i][2]!='6' && argv[i][2]!='x'))) {
	    if (strcmp(argv[i],"-fpic") && strcmp(argv[i],"-fno-pic"))
	      continue;
	  }
	*dest++=argv[i];
      }
#ifndef __DYN_LIB
      if (compile || _link) {
	*dest++="-isystem";
	*dest++=a;
      }
#else
      if (compile || _link || shared) {
	*dest++="-isystem";
	*dest++=a;
      }
#endif
      *dest++="-D__dietlibc__";
      if (mangleopts) {
	const char **o=Os;

	{
	  int fd;
	  tmp=getenv("HOME");
	  if (tmp) {
	    if (strlen(tmp)+strlen(cc)<900) {
	      strcpy(manglebuf,tmp);
	      strcat(manglebuf,"/.diet/");
	      tmp=strrchr(cc,'/');
	      if (tmp) ++tmp; else tmp=cc;
	      strcat(manglebuf,tmp);
	      if ((fd=open(manglebuf,O_RDONLY))>=0) {
		int len=read(fd,manglebuf,1023);
		if (len>0) {
		  manglebuf[len]=0;
		  *dest++=manglebuf;
		  for (i=1; i<len; ++i) {
		    if (manglebuf[i]==' ' || manglebuf[i]=='\n') {
		      manglebuf[i]=0;
		      if (i+1<len)
			*dest++=manglebuf+i+1;
		    }
		  }
		  goto incorporated;
		}
	      }
	    }
	  }
	}
	for (o=Os;*o;++o) {
	  if (!strcmp(*o,shortplatform)) {
	    ++o;
	    while (*o) {
	      *dest++=(char*)*o;
	      ++o;
	    }
	    break;
	  } else
	    while (*o) ++o;
	}
      }
incorporated:
      if (_link) {
	if (profile) *dest++="-lgmon";
	*dest++=c; *dest++=(char*)libgcc;
	if (!strcmp(shortplatform,"sparc") ||
	    !strcmp(shortplatform,"sparc64") ||
	    !strncmp(shortplatform,"arm",3)) {
	  *dest++=c;
	}
      }
#ifdef WANT_DYNAMIC
      if (_link) { *dest++=e; }
#endif
#ifdef __DYN_LIB
      if (shared){ *dest++=c; }
      f=alloca(strlen(platform)+100);
      if (_link) {
	strcpy(f,"-Wl,-dynamic-linker=");
	strcat(f,platform);
//	strcat(f,"/diet-linux.so");
	strcat(f,"/libdl.so");
	*dest++=f;
      }
#endif
      *dest=0;
      if (verbose) {
	for (i=0; newargv[i]; i++) {
	  __write2(newargv[i]);
	  __write2(" ");
	}
	__write2("\n");
      }
      execvp(newargv[0],newargv);
      goto error;
    }
  }
donttouch:
  execvp(argv[1],argv+1);
error:
  error("execvp() failed!\n");
  return 1;
}
